src/blob_format.h

```diff
@@ -291,6 +299,18 @@ class BlobFileMeta {
   std::string smallest_key_;
   std::string largest_key_;
 
+  // The effective size of current file. This is different from `file_size_`, as
+  // `file_size_` is the original size of the file, and does not consider space
+  // reclaimed by punch hole GC.
+  // We can't use file system's `st_blocks` to get the logical size, because
```

Comment:
I think it's okay to get the size as `effective_file_size` after restart. The size doesn't have to be so precise. Indeed, it may have false positive for triggering punch hole GC, but it would be updated to the accurate number after the gc scan. 
Then, we can get rid of updating manifest for `effective_file_size`